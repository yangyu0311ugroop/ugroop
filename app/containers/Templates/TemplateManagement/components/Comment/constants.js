export const COMMENT_STATUS = { RESOLVED: 'resolved', OPEN: 'open' };
