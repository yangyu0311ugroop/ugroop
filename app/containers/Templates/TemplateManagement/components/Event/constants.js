/**
 * Created by stephenkarpinskyj on 10/4/18.
 */

export const EVENT_ICON_SIZE = 40;
export const EVENT_ICON_SIZE_XS = 32;
export const EVENT_ICON_MARGIN = 4;
export const EVENT_ICON_SIZE_WITH_LABEL = 120;
export const EVENT_ICON_CUT_POINT = 70;
