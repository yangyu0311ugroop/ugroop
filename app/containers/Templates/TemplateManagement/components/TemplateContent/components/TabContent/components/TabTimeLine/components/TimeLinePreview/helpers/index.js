const FirstElement = 0;

function convertToArrayLists(data) {
  const lists = [];
  let index = 1;
  if (data && data.length > 0) {
    // has day 0
    const firstNode = data[FirstElement];
    lists.push({ id: firstNode.id, title: `Day ${index}` });
    let hasNextChildren = firstNode.nextNodes;
    while (hasNextChildren.length > 0) {
      index += 1;
      lists.push({
        id: hasNextChildren[FirstElement].id,
        title: `Day ${index}`,
      });
      hasNextChildren = hasNextChildren[FirstElement].nextNodes;
    }
  }
  return lists;
}

function traverseToLastNode(data) {
  let node = data[FirstElement];
  while (node && node.nextNodes.length > 0) {
    node = node.nextNodes[0];
  }
  return node;
}

export default { convertToArrayLists, traverseToLastNode };
