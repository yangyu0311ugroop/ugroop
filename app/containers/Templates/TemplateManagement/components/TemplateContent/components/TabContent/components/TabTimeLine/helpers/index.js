import _ from 'lodash';

const updateState = (state, filter, keys, values) => {
  // TODO: optimise the code to use deep clone.
  const cloneState = _.cloneDeep(state);
  const index = _.findIndex(cloneState, filter);
  let i = 0;
  if (index !== -1) {
    keys.forEach(key => {
      cloneState[index][key] = values[i];
      i += 1;
    });
  }
  return { cloneState, index };
};

export { updateState };
