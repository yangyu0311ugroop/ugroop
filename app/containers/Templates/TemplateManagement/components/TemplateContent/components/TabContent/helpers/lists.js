// TODO: Why is this file basically a duplicate of datastore/templateManagementStore/helper/lists.js????

import { List } from 'immutable';

const isOpposite = (lastOperation, newOperation) =>
  (lastOperation.action === 'MOVE_AFTER' &&
    newOperation.action === 'MOVE_BEFORE') ||
  (lastOperation.action === 'MOVE_BEFORE' &&
    newOperation.action === 'MOVE_AFTER');

const reverseAction = action =>
  action === 'MOVE_AFTER' ? 'MOVE_BEFORE' : 'MOVE_AFTER';

const isAcrossMove = ({ parentId, toBeMovedParentId }) =>
  toBeMovedParentId && parentId !== toBeMovedParentId;

const isEmpty = list => !List.isList(list) || list.size === 0;

const hasBeenMoved = (list, id) =>
  list.findIndex(({ toBeMovedId }) => toBeMovedId === id) !== -1;

const hasBeenMovedAcross = (list, id) =>
  list.findIndex(
    ({ toBeMovedId, parentId, toBeMovedParentId }) =>
      toBeMovedId === id && toBeMovedParentId && parentId !== toBeMovedParentId,
  ) !== -1;

export const canCancel = (lastOperation, newOperation) => {
  if (isAcrossMove(lastOperation) !== isAcrossMove(newOperation)) return false;

  // move 1 to after 2, then move 2 to after 1, nothing changed so we can cancel
  if (
    lastOperation.id === newOperation.toBeMovedId &&
    lastOperation.toBeMovedId === newOperation.id &&
    lastOperation.action === newOperation.action
  ) {
    return true;
  }

  // move 1 to after 2, then move 1 to before 2
  return (
    lastOperation.id === newOperation.id &&
    lastOperation.toBeMovedId === newOperation.toBeMovedId &&
    isOpposite(lastOperation, newOperation)
  );
};

// toBeMovedId and action must match
export const canOverride = (lastOperation, newOperation) => {
  if (!lastOperation || !lastOperation.id) return false;

  return (
    lastOperation.action === newOperation.action &&
    lastOperation.toBeMovedId === newOperation.toBeMovedId
  );
};

// check if we can reverse the actions
// i.e instead of move 1 to after 2, we move 2 to before 1.
// Result is same, but sometimes we can save 1 operation
const canReverse = (lastOperation, newOperation) =>
  isOpposite(lastOperation, newOperation) &&
  lastOperation.toBeMovedId === newOperation.id &&
  lastOperation.parentId === newOperation.parentId;

const reverse = (lastOperation, newOperation) => ({
  ...newOperation,
  id: newOperation.toBeMovedId,
  toBeMovedId: lastOperation.toBeMovedId,
  action: lastOperation.action,
});

// if operation is not moved across, toBeMoved is not dirty, node at position is dirty, it's better swap
const betterSwap = (list, operation) =>
  !isAcrossMove(operation) &&
  !hasBeenMoved(list, operation.toBeMovedId) &&
  hasBeenMoved(list, operation.id);

const swap = ({ toBeMovedId, id, action, ...operation }) => ({
  ...operation,
  id: toBeMovedId,
  toBeMovedId: id,
  action: reverseAction(action),
});

const push = params => list => {
  // if list is undefined or not a List
  if (!List.isList(list)) return List([params]);

  // if list is empty, simply push new action
  if (list.size === 0) return list.push(params);

  const nextOperation = params;

  const last = list.last();
  // 2 actions cancel each other, simply pop the last action out the pending queue
  if (canCancel(last, nextOperation)) {
    return list.pop();
  }

  // can reverse the action and node
  // i.e instead of move 1 to after 2, we move 2 to before 1.
  // Result is same, but sometimes we can save 1 operation
  if (canReverse(last, nextOperation)) {
    return list.push(reverse(last, nextOperation));
  }

  // otherwise push new action to end of list
  return list.push(nextOperation);
};

const defaultList = () => List();

const notBelongToSameParent = (operation = {}) => (node = {}) =>
  node.parentId !== operation.parentId;
// clear operations belong to same parent
const clear = operation => list =>
  (list || defaultList()).filter(notBelongToSameParent(operation));

const isFirstLevel = (node = {}) => !node.parentId;
// clear second level sections
const clearSections = list => (list || defaultList()).filter(isFirstLevel);

// discard all operations
const clearAll = () => list => (list || defaultList()).clear();

const toArray = list => {
  // if list is undefined or not a List
  if (!List.isList(list)) return [];

  return list.toArray();
};

// 2 adjacent actions can be merged when
// i.e. move 1 to 2 and then move 1 to 3 > move 1 to 3 directly
const optimiseList = list => {
  if (isEmpty(list)) return list;

  let currentList = list;
  let optimisedList = new List();

  while (currentList.size) {
    const lastOperation = optimisedList.last();
    const next = currentList.first();
    let nextOperation = next;

    // if last operation is dirty and this operation is not dirty, it's better swap
    if (betterSwap(optimisedList, next)) {
      nextOperation = swap(next);
    }

    if (canOverride(lastOperation, nextOperation)) {
      optimisedList = optimisedList.pop();
    }

    currentList = currentList.shift();
    optimisedList = push(nextOperation)(optimisedList);
  }

  return optimisedList;
};

const optimise = list => {
  const groups = list.groupBy(({ parentId }) =>
    parentId ? 'root' : 'children',
  );

  return groups.map(optimiseList).reduce((acc, values) => {
    let newAcc = acc;
    values.forEach(value => {
      newAcc = newAcc.push(value);
    });
    return newAcc;
  }, List());
};

export default {
  has: hasBeenMoved,
  hasBeenMoved,
  hasBeenMovedAcross,
  canReverse,
  reverseAction,
  betterSwap,
  swap,
  notBelongToSameParent,
  isFirstLevel,
  push,
  clear,
  clearSections,
  clearAll,
  toArray,
  isEmpty,
  optimise,
  optimiseList,
  default: defaultList,
};
