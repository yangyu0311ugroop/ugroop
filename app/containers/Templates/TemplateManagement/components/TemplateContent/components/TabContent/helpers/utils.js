import { findIndex, merge as lodashMerge, slice } from 'lodash';

const isNotMoving = ({ id, toBeMovedId } = {}) => id === toBeMovedId;

const isMovingAcross = ({ parentId, toBeMovedParentId } = {}) =>
  parentId !== toBeMovedParentId;

const isMovingInside = ({ id, action } = {}) => !id && action === 'MOVE_INSIDE';

const removeAt = (array = [], position) => [
  ...slice(array, 0, position),
  ...slice(array, position + 1),
];

const replaceAt = (array = [], position, value) => [
  ...slice(array, 0, position),
  value,
  ...slice(array, position + 1),
];

const insertTo = (array = [], position = array.length, value) => [
  ...slice(array, 0, position),
  value,
  ...slice(array, position),
];

const merge = (payload, ...others) => {
  let photos = {};
  // check if image is uploaded in custom data
  if (payload && payload.customData && payload.customData.photo) {
    photos = { photos: [{ content: payload.customData.photo }] };
  }

  return lodashMerge({}, payload, photos, ...others);
};

const checkRoot = (array, predicate, action, node, offset = 0) => {
  const index = findIndex(array, predicate);
  if (index === -1) {
    // not found, do nothing
    return array;
  }

  return action(array, index + offset, node);
};

const checkChildren = (array, predicate, subAction, action) => {
  // find in sub-level
  const index = findIndex(array, predicate);
  if (index === -1) {
    // not found, do nothing
    return array;
  }

  const { children, ...origin } = array[index];

  // insert to origin.children then replace it in origin
  const newChildren = subAction(children);
  const updatedOrigin = { ...origin, children: newChildren };

  // replace origin with updated one
  return action(array, index, updatedOrigin);
};

// compare 2 nodes by their IDs
const equaliser = (array2 = []) => (node1, index) =>
  array2[index] && node1 ? node1.id !== array2[index].id : true;

export default {
  isNotMoving,
  isMovingAcross,
  isMovingInside,
  removeAt,
  replaceAt,
  insertTo,
  merge,
  checkRoot,
  checkChildren,
  equaliser,
};
