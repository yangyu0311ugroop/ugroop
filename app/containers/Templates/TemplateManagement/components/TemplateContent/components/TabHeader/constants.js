export const FIXED_TABS = ['Days & Events', 'Gallery'];
export const FIXED_TABS_OFFSET = FIXED_TABS.length;
