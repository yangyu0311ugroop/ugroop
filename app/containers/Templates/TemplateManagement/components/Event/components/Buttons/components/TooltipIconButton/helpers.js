import { EVENT_CONSTANTS } from 'utils/constants/events';

const showSubIconFn = (type, subtype, props = {}) => {
  const { showSubIcon } = props;

  if (!showSubIcon) return false;

  const exception = [EVENT_CONSTANTS.TRANSPORTATIONS.FLIGHT.type];

  return exception.indexOf(subtype) === -1;
};

export const EVENT_ICON_HELPERS = {
  showSubIconFn,
};
