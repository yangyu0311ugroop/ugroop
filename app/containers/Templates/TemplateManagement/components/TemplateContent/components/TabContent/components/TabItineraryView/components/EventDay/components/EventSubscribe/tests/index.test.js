import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import React from 'react';
import { NODE_POSITIONS } from 'utils/constants/nodes';
import { MOMENT_HELPERS } from 'utils/helpers/moment';
import { EventSubscribe } from '../index';

describe('<EventSubscribe />', () => {
  let rendered;
  let instance;

  const props = {
    id: 1,
    position: NODE_POSITIONS.end,
    updateParentEvents: jest.fn(),
    index: 1,
    renderSection: 'ceiling',
    startingUp: true,
  };

  beforeEach(() => {
    rendered = shallow(<EventSubscribe {...props} />);
    instance = rendered.instance();
    jest.clearAllMocks();
  });

  it('should exists', () => {
    expect(EventSubscribe).toBeDefined();
  });

  it('should render without exploding', () => {
    expect(rendered.length).toBe(1);
  });

  describe('componentDidUpdate', () => {
    it('calls props.updateParentEvents', () => {
      rendered.setProps({ endTime: null });
      instance.componentDidUpdate({ endTime: 'not null' });
      expect(instance.props.updateParentEvents).toBeCalledWith(
        expect.objectContaining({
          id: instance.props.id,
          position: instance.props.position,
        }),
        true,
      );
    });
  });

  describe('componentWillUnmount', () => {
    it('calls props.updateParentEvents', () => {
      instance.componentWillUnmount();
      expect(instance.props.updateParentEvents).toBeCalledWith(
        expect.objectContaining({
          id: instance.props.id,
          position: instance.props.position,
        }),
        true,
      );
    });
  });

  describe('makeEventTime', () => {
    it('makes times at least SHORTEST_ITINERARY_DURATION apart', () => {
      MOMENT_HELPERS.setTimeZone = MOMENT_HELPERS.createUtc;
      rendered.setProps({
        renderSection: 'content',
        startTime: '2018-01-01T12:00:00.000Z',
        endTime: '2018-01-01T12:01:00.000Z',
      });
      const result = instance.makeEventTime(instance.props);
      expect(result.startTime.toISOString()).toEqual(
        '2018-01-01T12:00:00.000Z',
      );
      expect(result.endTime.toISOString()).toEqual('2018-01-01T12:30:00.000Z');
    });
  });

  describe('makeEvent', () => {
    it('still matches snapshot without event', () => {
      expect(instance.makeEvent()).toMatchSnapshot();
    });
  });

  describe('render()', () => {
    it('should render correctly', () => {
      const snapshot = shallow(<div>{instance.render()}</div>);
      expect(toJSON(snapshot)).toMatchSnapshot();
    });
  });
});
