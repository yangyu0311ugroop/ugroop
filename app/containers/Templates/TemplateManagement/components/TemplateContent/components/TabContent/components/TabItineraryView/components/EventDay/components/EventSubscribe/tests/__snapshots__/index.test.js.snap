// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EventSubscribe /> makeEvent still matches snapshot without event 1`] = `
Object {
  "dataId": undefined,
  "endTime": undefined,
  "id": undefined,
  "position": undefined,
  "renderSection": undefined,
  "startTime": undefined,
  "subtype": undefined,
  "type": undefined,
}
`;

exports[`<EventSubscribe /> render() should render correctly 1`] = `<div />`;
